helix-term/src/ui/editor.rs

```diff
@@ -437,52 +423,145 @@ impl EditorView {
             .map(|range| range.cursor_line(text))
             .collect();
 
-        for (i, line) in (view.offset.row..(last_line + 1)).enumerate() {
-            use helix_core::diagnostic::Severity;
-            if let Some(diagnostic) = doc.diagnostics().iter().find(|d| d.line == line) {
-                surface.set_stringn(
-                    viewport.x,
-                    viewport.y + i as u16,
-                    "●",
-                    1,
-                    match diagnostic.severity {
-                        Some(Severity::Error) => error,
-                        Some(Severity::Warning) | None => warning,
-                        Some(Severity::Info) => info,
-                        Some(Severity::Hint) => hint,
-                    },
-                );
+        fn fixed_width_gutter(
```

Comment:
I can't think of a situation where I've seen function-scoped functions. I find it makes it really hard to read what's going on in `render_gutter`, as the logic is interrupted by over 100 lines of helper-function definition.